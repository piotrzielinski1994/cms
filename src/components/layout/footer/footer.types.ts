import { DataFromGlobalSlug, TypedLocale } from 'payload';

export interface FooterProps {
  locale: TypedLocale;
}

export interface FooterClientProps {
  data: DataFromGlobalSlug<'footer'>;
}
