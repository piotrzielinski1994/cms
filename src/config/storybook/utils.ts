import { ReactNode } from 'react';

const DEFAULT_VALUE = '__default__' as const;
const THUMBNAIL_ID = 'thumbnail';

const storyToUrlPath = (title: string) => {
  const parts = title.split('/');
  const story = parts[parts.length - 1];
  const titlePart = title.toLowerCase().replace(/\//g, '-').replace(/\s+/g, '-');
  const storyPart = story
    .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
    .replace(/([a-zA-Z])(\d+)/g, '$1-$2')
    .toLowerCase();
  return `${titlePart}--${storyPart}`;
};

const getFallback = <T extends ReactNode>(value: T, translation: string): T => {
  return value === DEFAULT_VALUE ? (translation as T) : value;
};

export { DEFAULT_VALUE, getFallback, storyToUrlPath, THUMBNAIL_ID };
