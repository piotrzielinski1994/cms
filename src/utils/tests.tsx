import { Providers } from '@/providers';
import { ComponentProps, ReactNode } from 'react';

const resolveServerComponent = async <T extends (props: unknown) => Promise<ReactNode>>(
  Component: T,
  props: Parameters<T>[0],
) => {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
};

const withProviders = (
  children: ReactNode,
  props: Partial<Omit<ComponentProps<typeof Providers>, 'children'>> = {},
) => {
  const overwrittenProps: ComponentProps<typeof Providers> = {
    locale: 'en',
    initialTheme: 'light',
    initialFontScale: 'base',
    initialCookiesConsent: true,
    ...props,
    children,
  };
  return <Providers {...overwrittenProps} />;
};

export { resolveServerComponent, withProviders };
