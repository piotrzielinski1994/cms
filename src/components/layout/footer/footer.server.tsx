import { Page } from '@/payload/payload.types';
import { getCachedGlobal } from '@/payload/utils/globals';
import { Locale } from 'next-intl';
import { Footer as FooterComponent } from './footer';

type FooterProps = {
  locale: Locale;
};

const Footer = async ({ locale }: FooterProps) => {
  const footerData = await getCachedGlobal('footer', locale)();
  const navItems = (footerData?.navItems ?? []).map((it) => ({
    id: it.id!,
    label: it.link.label,
    path: String((it.link.reference?.value as Page).path),
  }));

  return <FooterComponent items={navItems} />;
};

export { Footer };
