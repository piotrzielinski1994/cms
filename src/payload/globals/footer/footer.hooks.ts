import { GlobalRevalidationTag } from '@/payload/utils/globals';
import { Locale } from 'next-intl';
import { revalidateTag } from 'next/cache';
import type { GlobalAfterChangeHook } from 'payload';

const revalidateFooter: GlobalAfterChangeHook = ({ doc, req: { payload, context, locale } }) => {
  if (!context.disableRevalidate) {
    payload.logger.info(`Revalidating footer`);
    const typedLocale = locale as Locale;
    const tag: GlobalRevalidationTag = `global__${typedLocale}__footer`;
    revalidateTag(tag);
  }

  return doc;
};

export { revalidateFooter };
