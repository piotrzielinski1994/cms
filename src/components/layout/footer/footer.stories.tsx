import type { Meta, StoryObj } from '@storybook/react';
import { type ComponentProps } from 'react';
import { Footer as FooterComponent } from './footer';

type Args = ComponentProps<typeof FooterComponent> & {
  label?: string;
};

const meta: Meta<Args> = {
  component: FooterComponent,
  title: 'Layout/Footer',
  argTypes: {
    items: { control: 'object' },
  },
  args: {
    items: [
      { id: 'nav-item-1', label: 'Link 1', href: '/' },
      { id: 'nav-item-2', label: 'Link 2', href: '/' },
      { id: 'nav-item-3', label: 'Link 3', href: '/' },
    ],
  },
};

const Render = (args: Args) => {
  return <FooterComponent {...args} />;
};

const Footer: StoryObj<typeof FooterComponent> = { render: Render };

export { Footer };
export default meta;
