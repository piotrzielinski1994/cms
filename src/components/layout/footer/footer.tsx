import { Logo } from '@/_old/components/Logo/Logo';
import { getCachedGlobal } from '@/_old/utilities/getGlobals';
import { cn } from '@/_old/utilities/ui';
import Container from '@/components/layout/container/container';
import Section from '@/components/layout/section/section';
import { Link } from '@/payload/locale/routing';
import { FooterProps } from './footer.types';

export async function Footer({ locale }: FooterProps) {
  const footerData = await getCachedGlobal('footer', locale)();
  const navItems = (footerData?.navItems ?? []).map((it) => ({
    id: it.id,
    label: it.link.label,
    // @ts-expect-error
    path: `/${it.link.reference?.value.slug}`,
  }));

  return (
    <Section as="footer" className="mt-auto py-5 bg-accent">
      <Container className={cn('flex items-center')}>
        <Link className="flex items-center" href="/">
          <Logo />
        </Link>
        <nav className="flex-grow flex justify-end ">
          <ul className="contents">
            {navItems.map(({ id, label, path }) => {
              return (
                <li key={id} className="contents">
                  <Link className="p-2" href={path}>
                    {label}
                  </Link>
                </li>
              );
            })}
          </ul>
        </nav>
      </Container>
    </Section>
  );
}
