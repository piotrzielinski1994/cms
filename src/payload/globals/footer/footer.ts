import { AdminTranslations, translations } from '@/config/locales.config';
import { link } from '@/payload/fields/link';
import type { GlobalConfig } from 'payload';
import { revalidateFooter } from './footer.hooks';

const footer: GlobalConfig = {
  slug: 'footer',
  access: {
    read: () => true,
  },
  admin: {
    group: {
      en: translations.en.common.component.plural,
      pl: translations.pl.common.component.plural,
    },
  },
  label: ({ t }: { t: AdminTranslations }) => t('components:footer'),
  fields: [
    {
      name: 'navItems',
      type: 'array',
      fields: [
        link({
          appearances: false,
        }),
      ],
      maxRows: 6,
      admin: {
        initCollapsed: true,
        components: {
          RowLabel: '@/payload/components/row-label#RowLabel',
        },
      },
    },
  ],
  hooks: {
    afterChange: [revalidateFooter],
  },
};

export { footer };
