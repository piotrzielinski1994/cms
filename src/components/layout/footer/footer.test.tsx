import { resolveServerComponent, withProviders } from '@/utils/tests';
import { render, waitFor } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { axe } from 'vitest-axe';
import { Footer } from './footer';

vi.mock('@/payload/utils/globals', () => ({
  getCachedGlobal: vi.fn().mockReturnValue(
    vi.fn().mockResolvedValue({
      navItems: [
        { id: '1', link: { label: 'Home', reference: { value: { path: '/' } } } },
        { id: '2', link: { label: 'About', reference: { value: { path: '/about' } } } },
      ],
    }),
  ),
}));

describe('Footer', () => {
  it('should have no accessibility violations', async () => {
    const ServerFooter = await resolveServerComponent(Footer, { locale: 'en' });
    const { container } = await waitFor(() => render(withProviders(<ServerFooter />)));
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should match the snapshot', async () => {
    const ServerFooter = await resolveServerComponent(Footer, { locale: 'en' });
    const { container } = render(withProviders(<ServerFooter />));
    expect(container).toMatchSnapshot();
  });
});
