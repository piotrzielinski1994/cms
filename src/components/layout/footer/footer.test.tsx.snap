// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > should match the snapshot 1`] = `
<div>
  <footer
    class="px-4 sm:px-6 grid cms-section mt-auto py-5 bg-background1"
  >
    <div
      class="w-full max-w-screen-2xl mx-auto cms-container flex items-center"
    >
      <a
        aria-label="Logo"
        class="text-2xl"
        href="/"
      >
        [Svg @/icons/logo.svg]
      </a>
      <nav
        class="flex-grow flex justify-end "
      >
        <ul
          class="contents"
        >
          <li
            class="contents"
          >
            <a
              class="p-2"
              href="/"
            >
              Home
            </a>
          </li>
          <li
            class="contents"
          >
            <a
              class="p-2"
              href="/about"
            >
              About
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </footer>
</div>
`;
